### DISCUSSION TITLE
Default Sort

### DISCUSSION THREAD

**Discussion by kb6673** [rappasoft/laravel-livewire-tables#275] [Q&A]
Chances are that I am missing this part of the docs - is there a way to apply a default sort?

I have tried doing it in the `query()` however this causes the orderBy to be applied to all frontend sorting too.

**Comment by rappasoft** [comment]
There is not for that reason, I usually pass it in with the route:

```route('frontend.user.index', ['sorts' => ['created_at' => 'desc']])```

That way it's not part of the query.

**Reply by davidabadi** [reply]
I think he is asking how to choose a default sorting column. Right now by default it doesn't sort, just grabs the data as it is in the database.

**Reply by rappasoft** [reply]
Yes, this is how I do that, there's no way to do that from the query because it would be applied to everything and make the other sorting not work. This essentially 'clicks' the column header for you in the direction you want when you click the link.

**Reply by kb6673** [reply]
I didn't try this yet - I got stuck instantly. We've got a page with multiple tables on, so I wasn't sure how to pass a default sort to a particular table (or multiple tables)

**Reply by davidabadi** [reply]
What about setting the sorts variable initial value to ['created_at' => 'desc']? That would add ?sorts%5Bcreated_at%5D=desc to the url. Isn't there a way to not add that to the url?

**Comment by simple-hacker** [comment]
I added default sorting with my forked branch for work.

In `WithSorting.php` I added two new properties

```php
public string $defaultSortColumn = '';
public string $defaultSortColumnDirection = 'asc';
```

And then in the `applySorting` method I added

```php
if (! empty($this->defaultSortColumn) && empty($this->sorts)) {
    return $query->orderBy($this->defaultSortColumn, $this->defaultSortColumnDirection);
}
```

And then for each table you can choose the default sort column and direction

**Reply by rappasoft** [reply]
I'll check it out but how does this not affect sorting overall? I.e. if you wanted to remove this sort from the UI?

**Reply by simple-hacker** [reply]
I'll push my PR later today to see.  This is for when you do not have any sorts on.  Currently it just lists the items in the order they were entered, so id.  My PR will allow it to sort by name, created_at, whatever column you need.  Once a sort is applied the defaultSortColumn will not be appended to the query.

**Reply by rappasoft** [reply]
Oh, I didn't see the other half of the check. Ok feel free to make a PR I have no problem if it all works correctly. Thanks!

**Reply by bryanburchfield** [reply]
This approach doesn't work if you're using a custom sortable, as on a relationship. eg.

```
Column::make("Related Model Name", "relatedModel.name")
    ->sortable(function (Builder $query, $direction) {
        return $query->orderBy(RelatedModel::select('name')
        ->whereColumn('base_model.related_model_id', 'related_model.id'), $direction);
    })
```

As a hack, I added a public class variable $initialized=false and altered the query() function so that the table is sorted by two columns on first load:

```
public function query(): Builder
{
    if (!$this->initialized) {
        $this->sorts = [
            'relatedModel.name' => 'asc',
            'name' => 'asc',
        ];
    }

    $this->initialized = true;

    return BaseModel::query()
        ->with('relatedModel')
        ...
```
